### ISSUE TITLE
Support for short-circuit evaluation

### ISSUE THREAD

**Issue by mmzeynalli** [python/mypy#17253]
**Bug Report**

Mypy fails to evaluate variable type in short-circuit evaluation.

**To Reproduce**

```python
class PortfolioQueryBuilder:
    def __init__(self, user_username: Optional[str] = None) :
        self.user_username = user_username

    def build_query(self):
        '''rest of the code/handling here'''

class Token:
    def __init__(self, username: str):
        self.username = username

auth = Token('username')
# auth = None
builder = PortfolioQueryBuilder(user_username=auth and auth.username)
```

**Expected Behavior**

It should not give an error, as line `auth and auth.username` always returns either `None` or username string, because of short-circuit evaluation.

**Actual Behavior**

When I run mypy, I get this error:

```
error: Argument "user_username" to "PortfolioQueryBuilder" has incompatible type "Token | str | None"; expected "str | None"  [arg-type]
```
**Your Environment**

- Mypy version used: mypy 1.9.0 (compiled: yes)
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files):
```ini
[tool.mypy]
plugins = ["sqlalchemy.ext.mypy.plugin", "pydantic.mypy"]
exclude = [
    "^admin\\.py$"
]
```
- Python version used: 3.11.5

**Comment by JelleZijlstra** [comment]
Mypy is technically correct here; it could be a subclass of Token that overrides `__bool__` to return False.

A bit arcane but I think this sort of pattern is better avoided in type-safe code.

**Comment by erictraut** [comment]
Consistent with Jelle's insight, the behavior changes if you mark the `Token` class as `@final`.

**Comment by JelleZijlstra** [comment]
Indeed, thanks @erictraut: https://mypy-play.net/?mypy=latest&python=3.10&gist=c808a9a03bd113724835d66cccdd062e. With that I don't think there's anything left to do here; mypy's behavior is correct and `@final` can be used to work around this behavior if it's a problem.

**Comment by mmzeynalli** [comment]
I see, @final did indeed work. Thanks

**Comment by mmzeynalli** [comment]
Okay, I have a little different circumstance, which gives the same error:
```python
@final
class Token(BaseModel):
    username: str
    exp: Optional[int] = None
    user_type: int


OptionalAuthUser = Annotated[TokenPayload | None, Depends(get_optional_user_access)]


@router.get(
    '',
    summary='Get portfolio list',
    status_code=status.HTTP_200_OK,
)
async def get_all_profiles_portfolios(
    db: DbDependency,
    auth: OptionalAuthUser,
):
    builder = PortfolioQueryBuilder(user_username=auth and auth.username)
    '''rest of the code'''
```

How should I act here, considering Token if wrapped into Annotated?

**Comment by JelleZijlstra** [comment]
You use `TokenPayload` but don't show that definition in your example. I would recommend rewriting as `auth.username if auth is not None else None`.

**Comment by mmzeynalli** [comment]
My bad, fixed it

**Comment by KotlinIsland** [comment]
@mmzeynalli I don't see any error, maybe you could provide a more complete example?

```py
from typing import final, Optional, Annotated

from fastapi import Depends
from pydantic import BaseModel


@final
class Token(BaseModel):
    username: str
    exp: Optional[int] = None
    user_type: int

def get_optional_user_access(): ...

OptionalAuthUser = Annotated[Token | None, Depends(get_optional_user_access)]


async def get_all_profiles_portfolios(
        auth: OptionalAuthUser,
):
    builder: str | None = auth and auth.username
    '''rest of the code'''
```
```
> mypy test.py
Success: no issues found in 1 source file
```
